### PR TITLE
[SPARK-31384][SQL] Fix NPE in OptimizeSkewedJoin

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

1. Fix NPE in `OptimizeSkewedJoin`

2. prevent other potential NPE errors in AQE.

### Why are the changes needed?

When there's a `inputRDD` of a plan has 0 partition, rule `OptimizeSkewedJoin` can hit NPE error because this kind of RDD means a null `MapOutputStatistics` due to:

https://github.com/apache/spark/blob/d98df7626b8a88cb9a1fee4f454b19333a9f3ced/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala#L68-L69

Thus, we should take care of such NPE errors in other places too.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

Added a test.

### CODE REVIEW

**Comment by Ngone51** [comment]
cc @cloud-fan @maryannxue

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala [hunk 1]
```diff
@@ -161,10 +161,14 @@ case class ShuffleQueryStageExec(
       case _ =>
     }
   }
+}
 
-  def mapStats: MapOutputStatistics = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala:165]: Can we make this return `Option[MapOutputStatistics]` and document it? It should be easier to use than an exactor.

**Review by maryannxue — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala [hunk 2]
```diff
@@ -162,9 +162,11 @@ case class ShuffleQueryStageExec(
     }
   }
 
-  def mapStats: MapOutputStatistics = {
-    assert(resultOption.isDefined)
-    resultOption.get.asInstanceOf[MapOutputStatistics]
+  def mapStats: Option[MapOutputStatistics] = {
+    assert(resultOption.isDefined, "ShuffleQueryStageExec should already be ready")
+    val stats = resultOption.get.asInstanceOf[MapOutputStatistics]
+    // mapStats can be null when the input RDD of a plan has 0 partition
```
Comment (Ngone51) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala:168]: nit: `mapStats` -> `stats`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala (hunk 1 above)
Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala:165]: updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala [hunk 3]
```diff
@@ -54,14 +53,10 @@ case class CoalesceShufflePartitions(session: SparkSession) extends Rule[SparkPl
     if (!shuffleStages.forall(_.shuffle.canChangeNumPartitions)) {
       plan
     } else {
-      val shuffleMetrics = shuffleStages.map { stage =>
-        assert(stage.resultOption.isDefined, "ShuffleQueryStageExec should already be ready")
-        stage.resultOption.get.asInstanceOf[MapOutputStatistics]
-      }
+      // `ShuffleQueryStageExec` returns None MapOutputStatistics when the input RDD
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala:56]: nit: `ShuffleQueryStageExec#mapStats` returns None when ...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DemoteBroadcastHashJoin.scala [hunk 4]
```diff
@@ -29,11 +28,11 @@ import org.apache.spark.sql.internal.SQLConf
 case class DemoteBroadcastHashJoin(conf: SQLConf) extends Rule[LogicalPlan] {
 
   private def shouldDemote(plan: LogicalPlan): Boolean = plan match {
-    case LogicalQueryStage(_, stage: ShuffleQueryStageExec) if stage.resultOption.isDefined
-      && stage.resultOption.get != null =>
-      val mapOutputStatistics = stage.resultOption.get.asInstanceOf[MapOutputStatistics]
-      val partitionCnt = mapOutputStatistics.bytesByPartitionId.length
-      val nonZeroCnt = mapOutputStatistics.bytesByPartitionId.count(_ > 0)
+    case LogicalQueryStage(_, s: ShuffleQueryStageExec) if s.resultOption.isEmpty => false
+    case LogicalQueryStage(_, s: ShuffleQueryStageExec) if s.mapStats.isDefined =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DemoteBroadcastHashJoin.scala:32]: nit: we can still use one case `case LogicalQueryStage(_, s: ShuffleQueryStageExec) if s.resultOption.isDefined && s.mapStates.isDefined`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala:168]: can we make the comment as a method doc?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala:168]: e.g.
```
/**
 * Returns the .... If the shuffle map stage has no partition, this method returns None, as there is no map statistics.
 */
def mapStats: Option[MapOutputStatistics] = {
  assert(resultOption.isDefined, "ShuffleQueryStageExec should already be ready")
  Option(resultOption.get.asInstanceOf[MapOutputStatistics])
}
```

**Comment by cloud-fan** [comment]
LGTM except one comment: https://github.com/apache/spark/pull/28153/files#r406112190

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderExec.scala [hunk 5]
```diff
@@ -99,12 +99,17 @@ case class CustomShuffleReaderExec private(
...
     val minSize = SQLMetrics.createSizeMetric(sparkContext, "minimum partition data size")
     val avgSize = SQLMetrics.createSizeMetric(sparkContext, "average partition data size")
-    val mapStats = shuffleStage.get.mapStats.bytesByPartitionId
-    val sizes = partitionSpecs.map {
-      case CoalescedPartitionSpec(startReducerIndex, endReducerIndex) =>
-        startReducerIndex.until(endReducerIndex).map(mapStats(_)).sum
-      case p: PartialReducerPartitionSpec => p.dataSize
-      case p => throw new IllegalStateException("unexpected " + p)
+    val mapStats = shuffleStage.get.mapStats
+    val sizes = if (mapStats.isDefined) {
```
Comment (maryannxue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderExec.scala:103]: nit: mapStats.map(...).getOrElse(Seq(0L)) ?

Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderExec.scala:103]: sounds good!

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by Ngone51** [comment]
thanks all!

**Pull request closed without merging**